### ISSUE TITLE
No validation dataset & data leak Problem

### ISSUE THREAD

**Issue by cindy-hong** [PhucNguyenAH/EEG-TCNTransformer#1]
Hey! I noticed that there is **no validation set** in your TCNTransformer.py file. However, the model that performs best on the **test set** is saved during training. So I'm wondering if the 83.41% accuracy of your model on the BCIIV 2a dataset was achieved this way? Because, based on your code, it appears that a **data leak** has occurred.

**Comment by PhucNguyenAH** [comment]
Thanks for your comment, I understand your raised issue. Train set and test set are already separated from the dataset BCIIV2a. In the validation process, the model is called eval()-evaluation function from PyTorch, so the model weight does not learn from the test set. I reference from the EEG-Conformer pipeline https://github.com/eeyhsong/EEG-Conformer.git

**Comment by cindy-hong** [comment]
Thanks for the reminder! I reviewed your code file as well as the EEG-Conformer code file. I'm wondering if the 83.41% accuracy of your model on the BCIIV2a dataset was calculated using the accuracy of the last epoch of each subject, or the average accuracy of each subject, or the best accuracy of each subject? For example, `.... Epoch: 1998   Train loss: 0.066274   Test loss: 1.835294   Train accuracy 0.979167   Test accuracy is 0.791667 Epoch: 1999   Train loss: 0.001446   Test loss: 2.097695   Train accuracy 1.000000   Test accuracy is 0.770833 The average accuracy is: 0.7983940972222211 The best accuracy is: 0.8923611111111112`

**Comment by PhucNguyenAH** [comment]
I select the best accuracy from the average accuracy of 9 subjects, and it can be any epoch. You can see more details in Table A5 in my paper https://doi.org/10.3390/signals5030034

**Comment by cindy-hong** [comment]
I see what you mean. My confusion is that when you record the accuracy of a particular subject, are you using the average accuracy of the trained model over all epochs? Or is it the best accuracy as written in your code? 

Like for example, if you got  `.... Epoch: 1998   Train loss: 0.066274   Test loss: 1.835294   Train accuracy 0.979167   Test accuracy is 0.791667 Epoch: 1999   Train loss: 0.001446   Test loss: 2.097695   Train accuracy 1.000000   Test accuracy is 0.770833 The average accuracy is: 0.7983940972222211 The best accuracy is: 0.8923611111111112` as a result, which is the accuracy of your model on this subject? `0.770833` or `0.798394` or `0.892361`?

**Comment by PhucNguyenAH** [comment]
I understand your confusion. You can see in the main(), it trains each subject independently. So these accuracies are related to 1 subject. For example, we train subject 5. The averAcc is the average of validation accuracy of subject 5 from epoch 1 to the current epoch. The bestAcc is the highest accuracy of subject 5 from epoch 1 to the current epoch. The testAcc is only the accuracy in the current epoch of subject 5. It is quite confusing because all of them are for 1 subject. Finally, in the main(), I got all highest accuracies of 9 subjects to calculate the final average highest accuracy. All highest accuracies of 9 subjects are not from the same epoch number. For example, the highest accuracy of subject 1 can be from epoch 4123, while the highest accuracy of subject 3 can be from epoch 4567.
